### PR TITLE
Remove more old code of projection analysis

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Addition to #55112.

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit fb666e137c86339b532f9becc956eb4ef60971d3 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55579/fb666e137c86339b532f9becc956eb4ef60971d3/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6502012735">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55579/fb666e137c86339b532f9becc956eb4ef60971d3/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Interpreters/SelectQueryOptions.h [hunk 1]
```diff
@@ -33,14 +33,6 @@ struct SelectQueryOptions
...
     bool ignore_limits = false;
-    /// This flag is needed to analyze query ignoring table projections.
-    /// It is needed because we build another one InterpreterSelectQuery while analyzing projections.
-    /// It helps to avoid infinite recursion.
-    bool ignore_projections = false;
-    /// This flag is also used for projection analysis.
-    /// It is needed because lazy normal projections require special planning in FetchColumns stage, such as adding WHERE transform.
-    /// It is also used to avoid adding aggregating step when aggregate projection is chosen.
-    bool is_projection_query = false;
     /// This flag is needed for projection description.
```
Comment (Algunenano) [inline src/Interpreters/SelectQueryOptions.h:36]: Should we update the comments around `ignore_ast_optimizations` and `ignore_setting_constraints` related to projections?

**Review by Algunenano — COMMENTED** [review]
I've left some minor comments about things I've seen in the changes.

The test failures seem related:
* `ClickHouse Stateless Tests (release, Analyzer)` - > Analyzer + projections + prewhere (`02796_projection_date_filter_on_view` and `01710_aggregate_projections`)
* `ClickHouse Stateless Tests (tsan, S3 Storage)`. TSAN detected a  couple of issues but they look related to CH local and not to these changes, but it should be confirmed.

**Comment by CurtizJ** [comment]
> The test failures seem related

Yes, they are related. I tried to fix it but unsuccessfully currently.

File: src/Interpreters/SelectQueryOptions.h (hunk 1 above)
Comment (CurtizJ) [reply src/Interpreters/SelectQueryOptions.h:36]: I think no, because we still use it in `ProjectionDescription`.

**Comment by CurtizJ** [comment]
Tests related to projections have been magically fixed by merging fresh master. I think it's because of #56502. So now we can merge this PR.

**Comment by Algunenano** [comment]
> So now we can merge this PR.

There is still some failing tests with the analyzer -> https://s3.amazonaws.com/clickhouse-test-reports/55579/69896d94bead717ebc75bdf1907935e6057ae98c/stateless_tests__release__analyzer_.html. Both involving projections and PREWHERE

**Comment by CurtizJ** [comment]
I reverted commit a01acf5d2a650fa7e7d4a2a7426fbdc29c5142c5 and returned back old behaviour when we used `StorageSnapshot` with projection and when we have access to both metadata snapshots of table and projection. For some reason using only projection metadata doesn't work in `optimizePrewhere`. We can fix it and remove projection from `StorageSnapshot` in separate PR.

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 2]
```diff
@@ -2556,56 +2493,28 @@ void InterpreterSelectQuery::executeFetchColumns(QueryProcessingStage::Enum proc
 
         /// Create optimizer with prepared actions.
         /// Maybe we will need to calc input_order_info later, e.g. while reading from StorageMerge.
-        if ((optimize_read_in_order || optimize_aggregation_in_order)
-            && (!query_info.projection || query_info.projection->complete))
+        if (optimize_read_in_order || optimize_aggregation_in_order)
```
Comment (Algunenano) [inline src/Interpreters/InterpreterSelectQuery.cpp:2496]: This if now would probably be easier to read if it was:
```
    if (optimize_read_in_order)
         ...
    else if (optimize_aggregation_in_order)
         ...
```

As both options are separate there is no need of doing (if (a || b) { if (a) else ...)

**Review by Algunenano — APPROVED** [review]

**Comment by CurtizJ** [comment]
Integration tests: failed tests with keeper which are flaky now.

Performance Comparison: `parallel_index` - #58593

**Pull request merged**